Gems/SurfaceData/Code/Include/SurfaceData/MixedStackHeapAllocator.h

```diff
@@ -0,0 +1,195 @@
+        {
+            // If we're trying to resize our static buffer, allow it to succeed as long as the new size is within the total size of the
+            // static buffer. Otherwise, return 0 to fail the resize request.
+            if (ptr == m_lastStaticAllocation)
+            {
+                if (newSize <= (sizeof(Node) * NumNodes))
+                {
+                    return newSize;
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+
+            // Resize from the heap.
+            return AZ::AllocatorInstance<AZ::SystemAllocator>::Get().Resize(ptr, newSize);
+        }
+
+        auto max_size() const -> size_type
+        {
+            // Since we allow both stack and heap allocations, the max allocation size for this container is the heap's maximum.
+            return AZ::AllocatorInstance<AZ::SystemAllocator>::Get().GetMaxContiguousAllocationSize();
+        }
+
+        auto get_allocated_size() const -> size_type
+        {
+            // Always return the full size of our stack allocation, plus the total amount of heap allocations.
+            // While this doesn't seem like an accurate result, it's consistent with how AZStd::allocator behaves.
+            // We would need to do a significant amount of extra bookkeeping to provide an accurate number here, and it doesn't
+            // appear like anything uses this value, so the extra bookkeeping isn't currently warranted.
+            return (sizeof(Node) * NumNodes) + AZ::AllocatorInstance<AZ::SystemAllocator>::Get().NumAllocatedBytes();
+        }
+
+        bool is_lock_free()
+        {
+            return false;
+        }
+
+        bool is_stale_read_allowed()
+        {
+            return false;
+        }
+
+        bool is_delayed_recycling()
+        {
+            return false;
+        }
+
+    private:
+        const char* m_name;
+
+        // This will point to &m_data if the static allocation is currently in use, and nullptr if it isn't.
+        // Eventually, this could also be used to support multiple static allocations from the same buffer.
+        void* m_lastStaticAllocation{ nullptr };
+
+        // Stack-based storage that exists for the same lifetime as the data structure using this allocator.
+        typename AZStd::aligned_storage<sizeof(Node) * NumNodes, AZStd::alignment_of<Node>::value>::type m_data;
+    };
+
+    template<class Node, AZStd::size_t NumNodes>
+    bool operator==(
+        [[maybe_unused]] const mixed_stack_heap_allocator<Node, NumNodes>& a,
+        [[maybe_unused]] const mixed_stack_heap_allocator<Node, NumNodes>& b)
+    {
+        return true;
+    }
+
+    template<class Node, AZStd::size_t NumNodes>
+    bool operator!=(
+        [[maybe_unused]] const mixed_stack_heap_allocator<Node, NumNodes>& a,
+        [[maybe_unused]] const mixed_stack_heap_allocator<Node, NumNodes>& b)
+    {
+        return false;
+    }
```

Comment:
A bit confused why == always returns true, it's worth a comment at least.

Reply:
Whoops, it's actually a bug, thx! I mistakenly thought _all_ allocators were supposed to trivially return true and false regardless of having different debug names. However, it turns out that they should only compare as equal if they can interchangeably allocate and deallocate each other's memory. They're interchangeable for heap allocations, but not stack allocations, so they're not truly interchangeable at all. I need to make these checks more strict.